src/include/sof/audio/module_adapter/module/module_interface.h

```diff
@@ -67,6 +67,79 @@ struct output_stream_buffer {
 	uint32_t size; /* size of data in the buffer */
 };
 
+/**
+ * \struct module_endpoint_ops
+ * \brief Ops relevant only for the endpoint devices such as the host copier or DAI copier.
+ *	  Other modules should not implement these.
+ */
+struct module_endpoint_ops {
+	/**
+	 * Returns total data processed in number bytes.
+	 * @param dev Component device
+	 * @param stream_no Index of input/output stream
+	 * @param input Selects between input (true) or output (false) stream direction
+	 * @return total data processed if succeeded, 0 otherwise.
+	 */
+	uint64_t (*get_total_data_processed)(struct comp_dev *dev, uint32_t stream_no, bool input);
+	/**
+	 * Retrieves component rendering position.
+	 * @param dev Component device.
+	 * @param posn Receives reported position.
+	 */
+	int (*position)(struct comp_dev *dev, struct sof_ipc_stream_posn *posn);
+	/**
+	 * Configures timestamping in attached DAI.
+	 * @param dev Component device.
+	 *
+	 * Mandatory for components that allocate DAI.
+	 */
+	int (*dai_ts_config)(struct comp_dev *dev);
+
+	/**
+	 * Starts timestamping.
+	 * @param dev Component device.
+	 *
+	 * Mandatory for components that allocate DAI.
+	 */
+	int (*dai_ts_start)(struct comp_dev *dev);
+
+	/**
+	 * Stops timestamping.
+	 * @param dev Component device.
+	 *
+	 * Mandatory for components that allocate DAI.
+	 */
+	int (*dai_ts_stop)(struct comp_dev *dev);
+
+	/**
+	 * Gets timestamp.
+	 * @param dev Component device.
+	 * @param tsd Receives timestamp data.
+	 *
+	 * Mandatory for components that allocate DAI.
+	 */
+	int (*dai_ts_get)(struct comp_dev *dev, struct timestamp_data *tsd);
+
+	/**
+	 * Fetches hardware stream parameters.
+	 * @param dev Component device.
+	 * @param params Receives copy of stream parameters retrieved from
+	 *	DAI hw settings.
+	 * @param dir Stream direction (see enum sof_ipc_stream_direction).
+	 *
+	 * Mandatory for components that allocate DAI.
+	 */
+	int (*dai_get_hw_params)(struct comp_dev *dev,
+				 struct sof_ipc_stream_params *params, int dir);
+
+	/**
+	 * Triggers device state.
+	 * @param dev Component device.
+	 * @param cmd Trigger command.
+	 */
+	int (*trigger)(struct comp_dev *dev, int cmd);
```

Comment:
how about get_attribute ops ? It is used to get base_module_cfg in each module. Do you plan to support it by get_configuration ?

Reply:
it is already done in the module adapter @RanderWang